datafusion/core/src/physical_optimizer/enforce_sorting.rs

```diff
@@ -162,8 +162,18 @@ impl PhysicalOptimizerRule for EnforceSorting {
         // remove unnecessary sorts, and optimize sort-sensitive operators:
         let adjusted = plan_requirements.transform_up(&ensure_sorting)?.data;
         let new_plan = if config.optimizer.repartition_sorts {
-            let plan_with_coalesce_partitions =
+            let data_vec: Vec<_> = adjusted.children.iter().map(|x| x.data).collect();
+            let mut plan_with_coalesce_partitions =
                 PlanWithCorrespondingCoalescePartitions::new_default(adjusted.plan);
+            plan_with_coalesce_partitions.children = plan_with_coalesce_partitions
+                .children
+                .into_iter()
+                .enumerate()
+                .map(|(id, mut child)| {
+                    child.data = data_vec[id];
+                    child
+                })
+                .collect();
```

Comment:
I think these changes are still unnecessary for this PR. Other than this, this PR is ready to merge. Thanks @Lordworms for this Pr.